src/TreeSelect.tsx

```diff
@@ -57,7 +57,7 @@ export default class TreeSelect extends React.Component<Props, State> {
   };
 
   public componentDidUpdate(_: Props, prevState: State) {
-    if (this.state.treeVisible && !prevState.treeVisible && this.filterInput)
+    if (this.state.treeVisible && this.filterInput && this.state.filter !== prevState.filter)
```

Comment:
`this.filterInput` is a safety condition for the `focus` method.

`this.state.treeVisible && !prevState.treeVisible` was a condition that checked that `treeVisible` changed. The updated code does not handle that case anymore, and incorrectly requires `treeVisible` to be true. Instead, consider this:

```js
const changesTriggeringFocus = [
  this.state.treeVisible !== prevState.treeVisible,
  this.state.filter !== prevState.filter,
];
if (this.filterInput && changesTriggeringFocus.indexOf(true) > -1) {
  this.filterInput.focus();
}
```

Reply:
`this.state.treeVisible && !prevState.treeVisible` checks a different thing than `this.state.treeVisible !== prevState.treeVisible`. The first checks if the component just gained focus. The second also checks if it just lost focus. If it just lost focus, we do not want to re-focus it.

If there's a better way to check if the component currently has focus than `this.state.treeVisible`, then I would happily use that. Otherwise, I think my logic is correct. "If the component is focused, and the filterInput exists, and the filter value just changed &rarr; focus filterInput".

Reply:
You’re right about the losing focus case. I think renaming `treeVisible` to `focused` will solve all problems.

Why does filter changing need to be now a part of the condition? The case of filter not changing, but `treeVisible` going from `false` to `true` does not trigger the focus action anymore.

Reply:
Ok, I changed that commit and added another. Let me know what you think.